File: oxidize-python/oxidize_python/core/model/dflash_weights.py
"""Weight containers mirroring oxidize-golang/core/model/dflash_weights.go."""

from __future__ import annotations

from dataclasses import dataclass

from oxidize_python.core.quantization import types as quant
from oxidize_python.core.quantization.dequant_k import dequantize
from oxidize_python.core.tensor import gemv


@dataclass
class QuantWeight:
    bytes: bytes
    q_type: quant.Type
    out_dim: int
    in_dim: int


@dataclass
class F32Weight:
    data: list[float]
    rows: int
    cols: int
    quant: QuantWeight | None = None

    @staticmethod
    def from_slice(data: list[float], rows: int, cols: int) -> F32Weight:
        return F32Weight(data=data, rows=rows, cols=cols)

    @staticmethod
    def from_quantized(raw: bytes, qtype: quant.Type, out_dim: int, in_dim: int) -> F32Weight:
        return F32Weight(
            data=[],
            rows=in_dim,
            cols=out_dim,
            quant=QuantWeight(bytes=raw, q_type=qtype, out_dim=out_dim, in_dim=in_dim),
        )

    def is_loaded(self) -> bool:
        return len(self.data) > 0 or self.quant is not None

    def _input_dim(self) -> int:
        if self.quant is not None:
            return self.quant.in_dim
        return self.cols

    def _output_dim(self) -> int:
        if self.quant is not None:
            return self.quant.out_dim
        return self.rows

    def _dequant_fn(self):
        if self.quant is None:
            return None
        qt = self.quant.q_type

        def _fn(row: bytes, out: list[float]) -> None:
            dequantize(qt, row, out)

        return _fn

    def gemv(self, input_: list[float], output: list[float]) -> None:
        if self.quant is not None:
            q = self.quant
            gemv.gemv_quantized_f32(
                q.bytes,
                self._dequant_fn(),
                q.out_dim,
                q.in_dim,
                input_,
                output,
                None,
            )
            return
        gemv.gemv_f32_transposed(self.data, self.cols, self.rows, input_, output)

    def gemm(self, inputs: list[float], outputs: list[float], batch: int) -> None:
        if batch <= 1:
            self.gemv(inputs, outputs)
            return
        if self.quant is not None:
            q = self.quant
            fn = self._dequant_fn()
            if fn is None:
                raise ValueError("missing dequant fn")
            gemv.gemm_quantized_f32(
                q.bytes,
                fn,
                q.out_dim,
                q.in_dim,
                1,
                batch,
                inputs,
                outputs,
                None,
            )
            return
        gemv.gemm_f32(inputs, self.data, batch, self.cols, self.rows, outputs)

    def row(self, row_idx: int, output: list[float]) -> None:
        if self.quant is not None:
            q = self.quant
            if row_idx >= q.out_dim or len(output) != q.in_dim:
                raise ValueError(
                    f"row bounds mismatch: row={row_idx} out={q.out_dim} want in={q.in_dim}"
                )
            row_bytes = quant.quantized_size(q.q_type, q.in_dim)
            start = row_idx * row_bytes
            end = start + row_bytes
            if end > len(q.bytes):
                raise ValueError("quantized row extends past tensor data")
            dequantize(q.q_type, q.bytes[start:end], output)
            return
        if row_idx >= self.rows or len(output) != self.cols:
            raise ValueError(
                f"row bounds mismatch: row={row_idx} rows={self.rows} "
                f"out={len(output)} cols={self.cols}"
            )
        base = row_idx * self.cols
        output[:] = self.data[base : base + self.cols]


def transpose_f32(data: list[float], gguf_rows: int, gguf_cols: int) -> list[float]:
    result = [0.0] * len(data)
    for r in range(gguf_rows):
        for c in range(gguf_cols):
            result[c * gguf_rows + r] = data[r * gguf_cols + c]
    return result


def quantized_gemv_supported(qtype: quant.Type, in_dim: int) -> bool:
    match qtype:
        case quant.Type.Q4_K_S | quant.Type.Q4_K_M | quant.Type.Q2_K | quant.Type.Q6_K:
            return in_dim % 256 == 0
        case quant.Type.Q8_0:
            return in_dim % 32 == 0
        case _:
            return False


def f32_weight_from_dims(data: list[float], dims: list[int]) -> F32Weight:
    if not dims:
        return F32Weight([], 0, 0)
    if len(dims) == 1:
        n = dims[0]
        return F32Weight.from_slice(data, n, 1)
    r, c = dims[0], dims[1]
    return F32Weight.from_slice(transpose_f32(data, r, c), c, r)


def weight_from_2d(data: list[float], rows: int, cols: int) -> F32Weight:
    if not data:
        return F32Weight([], 0, 0)
    if cols == 0:
        cols = len(data) // rows if rows else 0
    return F32Weight.from_slice(data, rows, cols)


def ones(n: int) -> list[float]:
    return [1.0] * n

File: oxidize-golang/core/model/dflash_weights.go
package model

import (
	"fmt"

	"github.com/Zapdev-labs/oxidize/golang/core/quantization"
	"github.com/Zapdev-labs/oxidize/golang/core/tensor"
)

// QuantWeight holds raw GGUF-quantized bytes in row-major out_dim × in_dim layout.
type QuantWeight struct {
	Bytes  []byte
	QType  quantization.Type
	OutDim int
	InDim  int
}

// F32Weight is a projection stored as f32 (transposed GEMV layout) or quantized bytes.
type F32Weight struct {
	Data  []float32
	Rows  int
	Cols  int
	Quant *QuantWeight
}

// NewF32WeightFromSlice builds a dense f32 weight (rows = output, cols = input).
func NewF32WeightFromSlice(data []float32, rows, cols int) F32Weight {
	return F32Weight{Data: data, Rows: rows, Cols: cols}
}

// NewF32WeightFromQuantized builds a weight backed by on-the-fly dequant GEMV.
func NewF32WeightFromQuantized(bytes []byte, qtype quantization.Type, outDim, inDim int) F32Weight {
	return F32Weight{
		Rows: inDim,
		Cols: outDim,
		Quant: &QuantWeight{
			Bytes:  bytes,
			QType:  qtype,
			OutDim: outDim,
			InDim:  inDim,
		},
	}
}

// IsLoaded reports whether the weight has f32 or quantized data.
func (w F32Weight) IsLoaded() bool {
	return len(w.Data) > 0 || w.Quant != nil
}

func (w F32Weight) inputDim() int {
	if w.Quant != nil {
		return w.Quant.InDim
	}
	return w.Cols
}

func (w F32Weight) outputDim() int {
	if w.Quant != nil {
		return w.Quant.OutDim
	}
	return w.Rows
}

func (w F32Weight) dequantFn() func([]byte, []float32) error {
	if w.Quant == nil {
		return nil
	}
	qt := w.Quant.QType
	return func(row []byte, out []float32) error {
		return quantization.DequantizeScalar(qt, row, out)
	}
}

// Gemv computes output = W * input.
func (w F32Weight) Gemv(input, output []float32) error {
	if w.Quant != nil {
		q := w.Quant
		// Fast path 1: C++ OXK fused kernels (liboxk).
		ok, err := quantization.GemvOxk(q.Bytes, q.QType, q.OutDim, q.InDim, input, output)
		if err != nil {
			return err
		}
		if ok {
			return nil
		}
		// Fast path 2: Rust AVX2+FMA kernel via CGo (oxidize-ffi).
		ok, err = quantization.GemvRust(q.Bytes, q.QType, q.OutDim, q.InDim, input, output)
		if err != nil {
			return err
		}
		if ok {
			return nil
		}
		// Fallback: pure-Go dequant+dot for any type native kernels don't handle.
		return tensor.GemvQuantizedF32(q.Bytes, w.dequantFn(), q.OutDim, q.InDim, input, output, nil)
	}
	return tensor.GemvF32Transposed(w.Data, w.Cols, w.Rows, input, output)
}

// Gemm computes batched matmul; batch=1 delegates to Gemv.
func (w F32Weight) Gemm(inputs, outputs []float32, batch int) error {
	if batch <= 1 {
		return w.Gemv(inputs, outputs)
	}
	if w.Quant != nil {
		q := w.Quant
		return tensor.GemmQuantizedF32(q.Bytes, w.dequantFn(), q.OutDim, q.InDim, 1, batch, inputs, outputs, nil)
	}
	return tensor.GemmF32(inputs, w.Data, batch, w.Cols, w.Rows, outputs)
}

// Row copies one output row (e.g. embedding vector) into output.
func (w F32Weight) Row(rowIdx int, output []float32) error {
	if w.Quant != nil {
		q := w.Quant
		if rowIdx >= q.OutDim || len(output) != q.InDim {
			return fmt.Errorf("row bounds mismatch: row=%d out=%d want in=%d", rowIdx, q.OutDim, q.InDim)
		}
		rowBytes, err := quantization.QuantizedSize(q.QType, q.InDim)
		if err != nil {
			return err
		}
		start := rowIdx * rowBytes
		end := start + rowBytes
		if end > len(q.Bytes) {
			return fmt.Errorf("quantized row extends past tensor data")
		}
		return quantization.DequantizeScalar(q.QType, q.Bytes[start:end], output)
	}
	if rowIdx >= w.Rows || len(output) != w.Cols {
		return fmt.Errorf("row bounds mismatch: row=%d rows=%d out=%d cols=%d", rowIdx, w.Rows, len(output), w.Cols)
	}
	copy(output, w.Data[rowIdx*w.Cols:(rowIdx+1)*w.Cols])
	return nil
}

func transposeF32(data []float32, ggufRows, ggufCols int) []float32 {
	result := make([]float32, len(data))
	for r := 0; r < ggufRows; r++ {
		for c := 0; c < ggufCols; c++ {
			result[c*ggufRows+r] = data[r*ggufCols+c]
		}
	}
	return result
}

func quantizedGemvSupported(qtype quantization.Type, inDim int) bool {
	switch qtype {
	case quantization.TypeQ4_K_S, quantization.TypeQ4_K_M, quantization.TypeQ2_K, quantization.TypeQ6_K:
		return inDim%256 == 0
	case quantization.TypeQ8_0:
		return inDim%32 == 0
	default:
		return false
	}
}
